contracts/DefaultStrategy.sol

```diff
@@ -522,248 +436,288 @@ contract DefaultStrategy is UUPSOwnableUpgradeable, UsingRegistryUpgradeable, Ma
+    }
 
-            if (remainingWithdrawal == 0) {
-                break;
-            }
+    /**
+     * @notice Adds/substracts value to totals of group and
+     * total stCELO in default strategy.
+     * @param group The validator group that we are updating.
+     * @param stCeloAmount The amount of stCELO.
+     * @param add Whether to add or substract.
+     */
+    function updateGroupStCelo(
+        address group,
+        uint256 stCeloAmount,
+        bool add
+    ) internal {
+        if (add) {
+            stCELOInGroup[group] += stCeloAmount;
+            totalStCeloInDefaultStrategy += stCeloAmount;
+        } else {
+            stCELOInGroup[group] -= stCeloAmount;
+            totalStCeloInDefaultStrategy -= stCeloAmount;
         }
+    }
 
-        return (
-            deprecatedGroupsWithdrawn,
-            deprecatedWithdrawalsPerGroup,
-            numberDeprecatedGroupsWithdrawn,
-            remainingWithdrawal
-        );
+    function getLesserAndGreaterOfActiveGroups(
+        address originalKey,
+        uint256 newValue,
+        bool withdrawal
+    ) internal view returns (address previous, address next) {
+        if (withdrawal) {
+            (previous, next) = getLesserAndGreaterOfActiveGroupsWithdrawal(originalKey, newValue);
+        } else {
+            (previous, next) = getLesserAndGreaterOfActiveGroupsDeposit(originalKey, newValue);
+        }
     }
 
     /**
-     * @notice Calculates how votes should be withdrawn from each active group.
-     * @param withdrawal The number of votes that need to be withdrawn.
-     * @return The array of group addresses that should be withdrawn from.
-     * @return The amount of votes to withdraw from the respective group in the
-     * array of groups withdrawn from.
+     * @notice Marks a group as deprecated.
+     * @param group The group to deprecate.
      */
-    function getActiveGroupWithdrawalDistribution(uint256 withdrawal)
-        internal
-        returns (address[] memory, uint256[] memory)
-    {
-        uint256 numberGroups = activeGroups.length();
-
-        if (withdrawal == 0 || numberGroups == 0) {
-            address[] memory noGroups = new address[](0);
-            uint256[] memory noWithdrawals = new uint256[](0);
-            return (noGroups, noWithdrawals);
+    function _deprecateGroup(address group) private {
+        if (!activeGroups.contains(group)) {
+            revert GroupNotActive(group);
+        }
+        activeGroups.remove(group);
+        unsortedGroups.remove(group);
+
+        uint256 groupTotalStCeloVotes = stCELOInGroup[group];
+
+        if (groupTotalStCeloVotes > 0) {
+            updateGroupStCelo(group, groupTotalStCeloVotes, false);
+            address[] memory fromGroups = new address[](1);
+            uint256[] memory fromVotes = new uint256[](1);
+            fromGroups[0] = group;
+            fromVotes[0] = IManager(manager).toCelo(groupTotalStCeloVotes);
```

Comment:
Do we need to call out to Manager to distribute within the strategy? Wouldn't it be possible to just use functions in this contract to do this correctly?

Reply:
In this case we do since we probably want to schedule real Celo transfer.

Reply:
I guess a call to to Manager might be still necessary to schedule the withdrawal and new deposits, but if I understand correctly, this calls Manager, which then calls back into here to get the vote distribution. Is there a clean way to avoid that call out and back in here?

Reply:
The distribution is being taken care of in DefaultStrategy on line 574 - _generateDepositVoteDistribution. Manager will only schedule transfer (scheduleTransferWithinStrategy) for already precomputed groups.